File: packages/module-sdk-node/src/proxy.ts
interface LlmMessage {
  role: string;
  content: string;
}

interface LlmCompleteOptions {
  provider: string;
  model: string;
  messages: LlmMessage[];
  schema?: Record<string, unknown>;
  maxOutputTokens?: number;
}

interface LlmCompleteResult {
  text: string;
  input_tokens?: number;
  output_tokens?: number;
}

async function call(operation: string, payload: Record<string, unknown>): Promise<Record<string, unknown>> {
  const base = process.env.RWB_PROXY_SOCKET;
  if (!base) throw new Error("RWB_PROXY_SOCKET is not configured");
  const url = `${base.replace(/\/$/, "")}/${operation}`;
  const response = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(120_000),
  });
  return (await response.json()) as Record<string, unknown>;
}

export async function llmComplete(options: LlmCompleteOptions): Promise<LlmCompleteResult> {
  const result = await call("llm.complete", {
    binding: { provider: options.provider, model_id: options.model },
    messages: options.messages,
    schema: options.schema ?? null,
    max_output_tokens: options.maxOutputTokens ?? null,
  });
  return result as unknown as LlmCompleteResult;
}

export async function journalNote(text: string, metadata?: Record<string, unknown>): Promise<Record<string, unknown>> {
  return call("journal.note", { text, metadata: metadata ?? {} });
}

export async function reviewRequest(payload: Record<string, unknown>, schema: Record<string, unknown>): Promise<Record<string, unknown>> {
  return call("review.request", { payload, schema });
}

export async function progressUpdate(percent: number, message: string): Promise<Record<string, unknown>> {
  return call("progress.update", { percent, message });
}

export async function artifactReadMetadata(artifactId: string): Promise<Record<string, unknown>> {
  return call("artifact.read_metadata", { artifact_id: artifactId });
}

File: packages/module-sdk-py/rwb_sdk/proxy.py
import json
import os
import urllib.request
from typing import Any


def _call(operation: str, payload: dict[str, Any]) -> dict[str, Any]:
    base = os.environ.get("RWB_PROXY_SOCKET")
    if not base:
        raise RuntimeError("RWB_PROXY_SOCKET is not configured")
    data = json.dumps(payload).encode("utf-8")
    request = urllib.request.Request(
        f"{base.rstrip('/')}/{operation}",
        data=data,
        method="POST",
        headers={"content-type": "application/json"},
    )
    with urllib.request.urlopen(request, timeout=120) as response:
        return json.loads(response.read().decode("utf-8"))


def llm_complete(provider: str, model: str, messages: list[dict[str, str]], schema: dict[str, Any] | None = None, max_output_tokens: int | None = None) -> dict[str, Any]:
    return _call(
        "llm.complete",
        {
            "binding": {"provider": provider, "model_id": model},
            "messages": messages,
            "schema": schema,
            "max_output_tokens": max_output_tokens,
        },
    )


def journal_note(text: str, metadata: dict[str, Any] | None = None) -> dict[str, Any]:
    return _call("journal.note", {"text": text, "metadata": metadata or {}})


def review_request(payload: dict[str, Any], schema: dict[str, Any]) -> dict[str, Any]:
    return _call("review.request", {"payload": payload, "schema": schema})


def progress_update(percent: int, message: str) -> dict[str, Any]:
    return _call("progress.update", {"percent": percent, "message": message})


def artifact_read_metadata(artifact_id: str) -> dict[str, Any]:
    return _call("artifact.read_metadata", {"artifact_id": artifact_id})
